[SPARK-22805][CORE] Use StorageLevel aliases in event logs

superbobry opened the pull request:
## What changes were proposed in this pull request?

The format of event logs uses a redundant representation for storage
levels, for instance, StorageLevel.DISK_ONLY is represented as

    {"Use Disk":true,"Use Memory":false,"Deserialized":false,"Replication":1}

which is 64 bytes more. This commit changes the event log representation
of the StorageLevel to predefined constants: NONE, DISK_ONLY, etc. The
change is fully backwards compatible.

## How was this patch tested?

`core` unit tests.

Commit:
[SPARK-22805][CORE] Use StorageLevel aliases in event logs

The format of event logs uses redundant representation for storage
levels, for instance StorageLevel.DISK_ONLY is represented as

    {"Use Disk":true,"Use Memory":false,"Deserialized":false,"Replication":1}

which is 64 bytes more. This commit changes the event log representation
of the StorageLevel to predefined constants: NONE, DISK_ONLY, etc. The
change is fully backward compatibly, because

* StorageLevel constructor is private, meaning that existing event
  logs can only contain these predefined levels;
* The JsonProtocol supports reading both the old format and the new one.

superbobry commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -444,11 +444,20 @@ private[spark] object JsonProtocol {
-    ("Deserialized" -> storageLevel.deserialized) ~
-    ("Replication" -> storageLevel.replication)
+  def storageLevelToJson(storageLevel: StorageLevel): JValue = storageLevel match {
+    case StorageLevel.NONE => "NONE"
```
Maybe we can change `StorageLevel.toString` to do this or add another method e.g. `StorageLevel.name`.

srowen replied:
Agree, that would be more robust.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -988,12 +997,16 @@ private[spark] object JsonProtocol {
-    val replication = (json \ "Replication").extract[Int]
-    StorageLevel(useDisk, useMemory, deserialized, replication)
+  def storageLevelFromJson(json: JValue): StorageLevel = json match {
+    case _: JString => StorageLevel.fromString(json.extract[String])
```
It's probably worth some comments about why there are two read paths

superbobry replied:
Done.

srowen commented on core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala:
```diff
@@ -2022,12 +1947,7 @@ private[spark] object JsonProtocolSuite extends Assertions {
       |      "Port": 300
       |    },
       |    "Block ID": "rdd_0_0",
-      |    "Storage Level": {
```
We'll also want to retain some tests of the old format to ensure it's still read. Maybe there are outside of the diff and I'm not seeing them.

superbobry replied:
Added a to/from JSON test for a custom `StorageLevel`.

srowen replied:
I guess I mean, doesn't this no longer test whether it can read the verbose, old style format? like this test does here and the ones above, that are being removed?

superbobry replied:
These strings are used with `testEvent` which assert that the serialized representation matches the one given in the string literal. See https://github.com/criteo-forks/spark/blob/7869e63a569a6fb6725996084f0c5c55fc130ac8/core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala#L457

squito replied:
I share sean's concern, I don't think your response is addressing it.  You added one test on L151 makes sure that some event which is *not* predefined still works.  But you don't have a test making sure that the old, verbose string can still be parsed (or is it somewhere else?)

probably this is indirectly covered by HistoryServerSuite but a more direrct test would be better

superbobry replied:
I've added a test ensuring all predefine storage levels can be read from the legacy format.

Sidenote: I've also noticed that the legacy format incorrectly handled the predefined `StorageLevel.OFF_HEAP` and an fact any other custom storage level with `useOffHeap = true`. It looks like a bug to me, wdyt?

squito replied:
yup, I completely agree that off heap is not respected in the json format.  can you file a bug?  I think its still relevant even after this goes in, for custom levels

srowen reviewed (changes requested):
@andrewor14 I think you touched this last, and a long time ago. Just pinging you in case you have an opinion.

Commit:
Addressed review comments

Note that the previous commit contained a bug -- user-defined storage
levels caused an exception in JsonProtocol.

vanzin commented on core/src/main/scala/org/apache/spark/storage/StorageLevel.scala:
```diff
@@ -132,6 +132,23 @@ class StorageLevel private(
   override def hashCode(): Int = toInt * 41 + replication
 
+  /** Name of the storage level if it is predefined or [[None]] otherwise. */
+  def name: Option[String] = this match {
```
`fromString` below has basically the opposite of this. How about storing the mapping in a `Seq[(StorageLevel, String)]` and using that in both methods? e.g. here it would be:

```
knownLevels.collect { case (level, name) if level == this => name }.headOption
```

And pretty similar code in `fromString`.

superbobry replied:
This sounds good, will do. A slightly unrelated point: I feel that the name `fromString` somehow implies that it's the opposite of `toString`. What do you think about renaming it to `fromName` now that we have `name`?

vanzin replied:
> What do you think about renaming 

It's a public method so renaming means breaking compatibility.

vanzin commented:
ok to test

SparkQA commented:
**[Test build #84984 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84984/testReport)** for PR 19992 at commit [`58e93bb`](https://github.com/apache/spark/commit/58e93bbec986d46007f6fd37abc6bdb419beb1b2).

SparkQA commented:
**[Test build #84984 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84984/testReport)** for PR 19992 at commit [`58e93bb`](https://github.com/apache/spark/commit/58e93bbec986d46007f6fd37abc6bdb419beb1b2).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/84984/
Test FAILed.

Commit:
Added StorageLevel.PREDEFINED to reduce boilerplate in name/fromString

SparkQA commented:
**[Test build #85006 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85006/testReport)** for PR 19992 at commit [`13fe385`](https://github.com/apache/spark/commit/13fe385f3139cbdf9c43ed5eec92cddcff7368f5).

SparkQA commented:
**[Test build #85006 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85006/testReport)** for PR 19992 at commit [`13fe385`](https://github.com/apache/spark/commit/13fe385f3139cbdf9c43ed5eec92cddcff7368f5).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85006/
Test FAILed.

Commit:
Fixed ScalaDoc

SparkQA commented:
**[Test build #85012 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85012/testReport)** for PR 19992 at commit [`e171f03`](https://github.com/apache/spark/commit/e171f03c27a11d0820f2ade22dc081005dc039a0).

superbobry commented:
Minor update: I've simulated #18162 on one of our 80G event logs and (unless there is a bug in the filtering code) the log shrank to 157M. The effect of this patch was almost negligible, it brought the size down to 155M. It is unclear for now if this generalizes to other workloads.

SparkQA commented:
**[Test build #85012 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85012/testReport)** for PR 19992 at commit [`e171f03`](https://github.com/apache/spark/commit/e171f03c27a11d0820f2ade22dc081005dc039a0).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85012/
Test FAILed.

Commit:
Fixed a typo in a comment

SparkQA commented:
**[Test build #85019 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85019/testReport)** for PR 19992 at commit [`81b980f`](https://github.com/apache/spark/commit/81b980fbee125f50013fe512ffe64480a8c16f88).

SparkQA commented:
**[Test build #85019 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85019/testReport)** for PR 19992 at commit [`81b980f`](https://github.com/apache/spark/commit/81b980fbee125f50013fe512ffe64480a8c16f88).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85019/
Test FAILed.

superbobry commented:
Can someone have a look at the tests, please? I can't see the failure (and in theory, the change should not affect SparkR).

srowen commented:
Ignore the error, it's being fixed separately.

SparkQA commented:
**[Test build #4016 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4016/testReport)** for PR 19992 at commit [`81b980f`](https://github.com/apache/spark/commit/81b980fbee125f50013fe512ffe64480a8c16f88).

SparkQA commented:
**[Test build #4016 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4016/testReport)** for PR 19992 at commit [`81b980f`](https://github.com/apache/spark/commit/81b980fbee125f50013fe512ffe64480a8c16f88).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

vanzin commented on core/src/main/scala/org/apache/spark/storage/StorageLevel.scala:
```diff
@@ -132,6 +132,10 @@ class StorageLevel private(
 
+  /** Name of the storage level if it is predefined or `None` otherwise. */
+  def name: Option[String] = StorageLevel.PREDEFINED
+    .collectFirst { case (storageLevel, name) if storageLevel == this => name }
```
If the method body needs multiple lines it's better to wrap it with `{ }`. Same thing below.

Commit:
Added braces to StorageLevel.{name,fromString}

SparkQA commented:
**[Test build #85135 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85135/testReport)** for PR 19992 at commit [`59ed873`](https://github.com/apache/spark/commit/59ed8731f12d87954be716fa3c0914cda3d3690f).

vanzin commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -444,12 +444,15 @@ private[spark] object JsonProtocol {
-    ("Deserialized" -> storageLevel.deserialized) ~
-    ("Replication" -> storageLevel.replication)
-  }
+  def storageLevelToJson(storageLevel: StorageLevel): JValue =
```
Braces.

superbobry replied:
Sorry, missed it. I've decided not to add braces to `storageLevelFromJson` because it seems to look OK with the toplevel match.

SparkQA commented:
**[Test build #85135 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85135/testReport)** for PR 19992 at commit [`59ed873`](https://github.com/apache/spark/commit/59ed8731f12d87954be716fa3c0914cda3d3690f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85135/
Test PASSed.

Commit:
More braces

jiangxb1987 commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -444,12 +444,15 @@ private[spark] object JsonProtocol {
-    ("Replication" -> storageLevel.replication)
-  }
+  def storageLevelToJson(storageLevel: StorageLevel): JValue =
+    storageLevel.name match {
```
nit:
```
storageLevel.name.getOrElse(...)
```

superbobry replied:
Done. Sadly, in this case, getOrElse requires explicit type.

superbobry replied:
Actually, after seeing the compilation error, I recall why I went with a match instead of `getOrElse` -- the former does not require an explicit conversion to JString.

jiangxb1987 commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -988,12 +991,20 @@ private[spark] object JsonProtocol {
+      val replication = (json \ "Replication").extract[Int]
+      StorageLevel(useDisk, useMemory, deserialized, replication)
+    case _ =>
+      throw new IllegalArgumentException(json.toString)
```
nit:
```
throw new IllegalArgumentException(s"Invalid storage level from json: $json.")
```

superbobry replied:
Done. I've changed the message to match the one in `accumValueFromJson`.

Commit:
Addressed more review comments

SparkQA commented:
**[Test build #85402 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85402/testReport)** for PR 19992 at commit [`1fc6e75`](https://github.com/apache/spark/commit/1fc6e7532f7a9542d18183334912ece4a4c4c9cc).

SparkQA commented:
**[Test build #85402 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85402/testReport)** for PR 19992 at commit [`1fc6e75`](https://github.com/apache/spark/commit/1fc6e7532f7a9542d18183334912ece4a4c4c9cc).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85402/
Test FAILed.

SparkQA commented:
**[Test build #85403 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85403/testReport)** for PR 19992 at commit [`b1e6f5f`](https://github.com/apache/spark/commit/b1e6f5f1503604ebf1e9e695c219dcfa11ae91b7).

jiangxb1987 commented:
LGTM

SparkQA commented:
**[Test build #85403 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85403/testReport)** for PR 19992 at commit [`b1e6f5f`](https://github.com/apache/spark/commit/b1e6f5f1503604ebf1e9e695c219dcfa11ae91b7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85403/
Test PASSed.

Commit:
Added a test explicitly ensuring backward compatbility of StorageLevel

SparkQA commented:
**[Test build #85579 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85579/testReport)** for PR 19992 at commit [`9fbfe40`](https://github.com/apache/spark/commit/9fbfe40f5ca83f080f56f3e91c7a6f3f27471df5).

SparkQA commented:
**[Test build #85579 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85579/testReport)** for PR 19992 at commit [`9fbfe40`](https://github.com/apache/spark/commit/9fbfe40f5ca83f080f56f3e91c7a6f3f27471df5).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85579/
Test FAILed.

SparkQA commented:
**[Test build #85580 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85580/testReport)** for PR 19992 at commit [`cb1fe6a`](https://github.com/apache/spark/commit/cb1fe6a572d8085d36884bf950a840f972976458).

SparkQA commented:
**[Test build #85580 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85580/testReport)** for PR 19992 at commit [`cb1fe6a`](https://github.com/apache/spark/commit/cb1fe6a572d8085d36884bf950a840f972976458).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85580/
Test PASSed.

squito commented:
change is fine, but from discussion on the jira I'm unclear if this is really worth it -- gain seems pretty small after the other fix in 2.3.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

superbobry commented:
@squito I think it's fine to just close the PR/JIRA issue.

squito commented:
thanks for looking into this @superbobry -- can you actually close this yourself?  we can't directly close it (there is a way but its more complicated)
